### DISCUSSION TITLE
How to specify multipart/form-data request body in path macro?

### DISCUSSION THREAD

**Discussion by mrf7777** [juhaku/utoipa#742] [Q&A]
I was looking at the latest path attribute macro documentation on docs.rs (utoipa-3.4.4) but could not find how to specify that a function could accept a request body with data with content type `multipart/form-data`.

For context, I am using utoipa to generate an OpenAPI spec to generate a typescript client to talk to my rust web server running actix-web. I intend to use `multipart/form-data` to upload a single file from the browser. I have considered trying to make an endpoint on my actix-web server to accept a raw file, but I struggle with annotating (with utoipa) the fact that multiple specific types of files can be uploaded and I have to resort to `application/octet-stream` to catch all file/content types and then use sniffing to deduce the content type uploaded. An endpoint that takes `multipart/form-data` would allow me to inspect the content type of the uploaded content easily.

**Comment by juhaku** [comment]
I belive it should be doable with something like this definition:
```rust
#[utoipa::path(
    request_body(content = WhatEverStruct, description = "Multipart file", content_type = "multipart/form-data"),
)]
```

**Reply by mrf7777** [reply]
And I guess that the `WhatEverStruct` will specify the expected keys of the form data passed to the function? I am using [actix-multipart](https://docs.rs/actix-multipart/latest/actix_multipart/) to represent that `content` struct but I think that there is an issue open for implementing this: #740. So, is there a manual way to specify the expected keys in the request body or elsewhere?

**Reply by juhaku** [reply]
The `WhatEverStruct` must implement `ToSchema` trait and so there are couple of options to choose from.
1. Is to use a general type, like `Value` or `Object`. These will render as "typeless" types e.g. `Object` will be the OpenAPI's `type: object` and `Value` is the JSON value indicating OpenAPI's "Any" type that can basically accept anything. With these of course then there will not be any fields generated in the docs but should make the compiler happy.
2. Would be to implement the `ToSchema` for the `WhatEverStruct` by the means possible e.g. using a "new type" pattern. More insight here: https://github.com/juhaku/utoipa/issues/656 and here https://github.com/juhaku/utoipa/issues/507#issuecomment-1465007714.
